File: lib/fileorganizer.py
from PyQt6.QtWidgets import QMainWindow, QApplication, QVBoxLayout
from PyQt6.QtGui import QPixmap
from PyQt6.uic import loadUi
from getpass import getuser
from shutil import move
from PyQt6 import QtGui
from sys import argv
from os import listdir, getcwd, rename, remove, path, mkdir

class FileOrganizer(QMainWindow):
    allFile = []
    txt = []
    png = []
    jpg = []
    mp4 = []
    mkv = []
    mov = []
    mp3 = []
    gif = []
    webp = []
    webm = []
    heic = []
    etc = []

    path = ''
	
    def __init__(self):
        super(FileOrganizer, self).__init__()
        loadUi('UI/Main.ui', self)
        
        self.setWindowIcon(QtGui.QIcon(getcwd() + '/logo/WHI3PER.png'))
        self.PATH.insert('C:/Users/' + getuser() + '/Desktop/')
        
        # ========================================================== #
        
        self.SHOW.clicked.connect(self.showDirectory)
        self.CATEGORIES.clicked.connect(self.categories)
        self.RENAME.clicked.connect(self.rename)
        self.DELETE.clicked.connect(self.delete)
        
    # ============================================================================================== #
    
    def checkDirectory(self):
        if path.exists(self.path):
            return True
        else:
            return False

    # ============================================================================================== #

    def clearVariables(self):
        self.allFile.clear()
        self.txt.clear()
        self.png.clear()
        self.jpg.clear()
        self.mp4.clear()
        self.mkv.clear()
        self.mov.clear()
        self.mp3.clear()
        self.gif.clear()
        self.webp.clear()
        self.webm.clear()
        self.heic.clear()
        self.etc.clear()
    
        path = ''
        
    # ============================================================================================== #
        
    def saveFileNames(self):
        self.clearVariables()
        self.path = self.PATH.text()
        
        if self.path[-2:] != '/':
            self.path = self.path + '/'

        if not self.checkDirectory():
            self.warning('None', 2)
        else:
            listDirectory = listdir(self.path)

            for item in listDirectory:
                if item[-4:] == '.cpp' or item[-4:] == '.txt' or item[-3:] == '.py':
                    self.txt.append(item)
                elif item[-4:] == '.png' or item[-4:] == '.PNG':
                    self.png.append(item)
                elif item[-4:] == '.jpg' or item[-4:] == '.JPG':
                    self.jpg.append(item)
                elif item[-4:] == '.mp4' or item[-4:] == '.MP4':
                    self.mp4.append(item)
                elif item[-4:] == '.mkv' or item[-4:] == '.MKV':
                    self.mkv.append(item)
                elif item[-4:] == '.mov' or item[-4:] == '.MOV':
                    self.mov.append(item)
                elif item[-4:] == '.mp3' or item[-4:] == '.MP3':
                    self.mp3.append(item)
                elif item[-4:] == '.gif' or item[-4:] == '.GIF':
                    self.gif.append(item)
                elif item[-5:] == '.webp' or item[-5:] == '.WEBP':
                    self.webp.append(item)
                elif item[-5:] == '.webm' or item[-5:] == '.WEBM':
                    self.webm.append(item)
                elif item[-5:] == '.heic' or item[-5:] == '.HEIC':
                    self.heic.append(item)
                else:
                    self.etc.append(item)

            self.allFile.append(self.txt)
            self.allFile.append(self.png)
            self.allFile.append(self.jpg)
            self.allFile.append(self.mp4)
            self.allFile.append(self.mkv)
            self.allFile.append(self.mov)
            self.allFile.append(self.mp3)
            self.allFile.append(self.gif)
            self.allFile.append(self.webp)
            self.allFile.append(self.webm)
            self.allFile.append(self.heic)
            self.allFile.append(self.etc)
        
    # ============================================================================================== #

    def showDirectory(self):
        self.LOGS.clear()
        self.LOGS.append('\t\t   *** Show Directory ***\n')
        
        self.saveFileNames()
        
        for type in self.allFile:
            if len(type) != 0:
                index = 1
                for file in type:
                    self.LOGS.append(f'[{index}/{len(type)}] -- {file}')
                    index += 1
                self.LOGS.append('')
    
    # ============================================================================================== #

    def categories(self):
        self.LOGS.clear()
        self.LOGS.append('\t\t    *** Categories ***\n')
        
        if not self.checkDirectory():
            self.warning('None', 2)
        else:
            if self.txt: self.gotoCategories(self.txt, 'TXT')
            if self.png: self.gotoCategories(self.png, 'PNG')
            if self.jpg: self.gotoCategories(self.jpg, 'JPG')
            if self.mp4: self.gotoCategories(self.mp4, 'MP4')
            if self.mkv: self.gotoCategories(self.mkv, 'MKV')
            if self.mov: self.gotoCategories(self.mov, 'MOV')
            if self.mp3: self.gotoCategories(self.mp3, 'MP3')
            if self.gif: self.gotoCategories(self.gif, 'GIF')
            if self.webp: self.gotoCategories(self.webp, 'WEBP')
            if self.webm: self.gotoCategories(self.webm, 'WEBM')
            if self.heic: self.gotoCategories(self.heic, 'HEIC')
            if self.etc: self.gotoCategories(self.etc, 'ETC')
    
    # ============================================================================================== #

    def gotoCategories(self, File, Format):
        self.LOGS.append('~ ' + Format)
        new_path = (self.path + '/' + Format + '/')
        
        if not path.exists(new_path):
            mkdir(new_path)
            
        for item in File:
            move((self.path + item), (new_path + item))

    # ============================================================================================== #
    
    def warning(self, Format: str, case: int):
        if case == 1:
            self.LOGS.append(f'Warning: There is no \'{Format}\' file or your request is invalid.')
        else:
            self.LOGS.append(f'Warning: This directory does not exist !')
        
    # ============================================================================================== #
        
    def rename(self):
        self.saveFileNames()
        self.LOGS.clear()
        self.LOGS.append('\t\t      *** Rename ***\n')

        if not self.checkDirectory():
            self.warning('None', 2)
        else:
            match(self.RENAME_COMBOBOX.currentText()):
                case 'TXT': self.gotoRename(self.txt, 'TXT', 4)
                case 'PNG': self.gotoRename(self.png, 'PNG', 4)
                case 'JPG': self.gotoRename(self.jpg, 'JPG', 4)
                case 'MP4': self.gotoRename(self.mp4, 'MP4', 4)
                case 'MKV': self.gotoRename(self.mkv, 'MKV', 4)
                case 'MOV': self.gotoRename(self.mov, 'MOV', 4)
                case 'MP3': self.gotoRename(self.mp3, 'MP3', 4)
                case 'GIF': self.gotoRename(self.gif, 'GIF', 4)
                case 'WEBP': self.gotoRename(self.webp, 'WEBP', 5)
                case 'WEBM': self.gotoRename(self.webm, 'WEBM', 5)
                case 'HEIC': self.gotoRename(self.heic, 'HEIC', 5)
        
    # ============================================================================================== #

    def gotoRename(self, File, Format, FormatLength):
        self.LOGS.append('~ ' + Format)
        number = 1

        if File:
            for item in File:
                if path.exists(self.path + str(number) + item[-FormatLength:]):
                    number += 1
                else:
                    self.LOGS.append(item)
                    rename((self.path + item), (self.path + str(number) + item[-FormatLength:]))
                    number += 1
                
            self.LOGS.append(f'\nResult: {len(File)} {Format} files have been successfully renamed.')
        else:
            self.warning(Format, 1)
        
    # ============================================================================================== #
        
    def delete(self):
        self.saveFileNames()
        self.LOGS.clear()
        self.LOGS.append('\t\t      *** Delete ***\n')
        
        if not self.checkDirectory():
            self.warning('None', 2)
        else:
            match(self.DELETE_COMBOBOX.currentText()):
                case 'ALL':
                    self.LOGS.append('~ ALL')
                    
                    for type in self.allFile:
                        if len(type) != 0:
                            for file in type:
                                self.LOGS.append(file)
                                remove(self.path + file)
                    self.LOGS.append('\nResult: All files have been successfully deleted.')
                
                case 'TXT': self.gotoDelete(self.txt, 'TXT')
                case 'PNG': self.gotoDelete(self.png, 'PNG')
                case 'JPG': self.gotoDelete(self.jpg, 'JPG')
                case 'MP4': self.gotoDelete(self.mp4, 'MP4')
                case 'MKV': self.gotoDelete(self.mkv, 'MKV')
                case 'MOV': self.gotoDelete(self.mov, 'MOV')
                case 'MP3': self.gotoDelete(self.mp3, 'MP3')
                case 'GIF': self.gotoDelete(self.gif, 'GIF')
                case 'WEBP': self.gotoDelete(self.webp, 'WEBP')
                case 'WEBM': self.gotoDelete(self.webm, 'WEBM')
                case 'HEIC': self.gotoDelete(self.heic, 'HEIC')
                case 'ETC': self.gotoDelete(self.etc, 'ETC')
        
    # ============================================================================================== #

    def gotoDelete(self, File, Format):
        self.LOGS.append('~ ' + Format)
        if File:
            for item in File:
                self.LOGS.append(item)
                remove(self.path + item)
            self.LOGS.append(f'\nResult: {len(File)} {Format} files have been successfully deleted.')
        else:
            self.warning(Format, 1)

    # ============================================================================================== #

File: fileorganizer.cpp
#include "fileorganizer.h"
#include "./ui_fileorganizer.h"

FileOrganizer::FileOrganizer(QWidget *parent): QMainWindow(parent), ui(new Ui::FileOrganizer) {
    ui->setupUi(this);
}

FileOrganizer::~FileOrganizer() {
    delete ui;
}

str FileOrganizer::GetPath() const {
    str text{ ui->Path->text().toUtf8().toStdString() };

    if (text.find('\\') < 100) {
        for (std::size_t i{ 0 }; i < text.length(); ++i) {
            if (text.at(i) == '\\') {
                text.replace(i, 1, "/");
            }
        }
    }

    if (!text.ends_with("/")) {
        text.append("/");
    }

    return text;
}

bool FileOrganizer::ChackExistsDirectory(str& path) {
    if (fs::exists(path) && fs::is_directory(path)) {
        return true;
    }

    else {
        return false;
    }
}

void FileOrganizer::ClearAllVectors() {
    this->Files.clear();
    // this->TXT.clear();
    // this->PNG.clear();
    // this->JPG.clear();
    // this->MP4.clear();
    // this->MKV.clear();
    // this->MOV.clear();
    // this->MP3.clear();
    // this->GIF.clear();
    // this->WEBP.clear();
    // this->WEBM.clear();
    // this->HEIC.clear();
    // this->ETC.clear();
}

bool FileOrganizer::SaveFilesNames() {
    this->ClearAllVectors();
    this->Path = this->GetPath();

    if (!this->ChackExistsDirectory(this->Path)) {
        this->Warning("None", 2);
        return false;
    }

    if (fs::is_empty(this->Path)) {
        this->Warning("None", 3);
        return false;
    }

    else {
        try {
            for (const auto& entry : fs::directory_iterator(this->Path)) {
                if (fs::is_regular_file(entry.path())) {
                    str FileName{ entry.path().string().substr(this->Path.size(), entry.path().string().length() - 1) };

                    if (FileName.ends_with("txt"))
                        this->Files["TXT"].push_back(FileName);
                    else if (FileName.ends_with("png") or FileName.ends_with("PNG"))
                        this->Files["PNG"].push_back(FileName);
                    else if (FileName.ends_with("jpg") or FileName.ends_with("JPG"))
                        this->Files["JPG"].push_back(FileName);
                    else if (FileName.ends_with("mp4") or FileName.ends_with("MP4"))
                        this->Files["MP4"].push_back(FileName);
                    else if (FileName.ends_with("mkv") or FileName.ends_with("MKV"))
                        this->Files["MKV"].push_back(FileName);
                    else if (FileName.ends_with("mov") or FileName.ends_with("MOV"))
                        this->Files["MOV"].push_back(FileName);
                    else if (FileName.ends_with("mp3") or FileName.ends_with("MP3"))
                        this->Files["MP3"].push_back(FileName);
                    else if (FileName.ends_with("gif") or FileName.ends_with("GIF"))
                        this->Files["GIF"].push_back(FileName);
                    else if (FileName.ends_with("webp") or FileName.ends_with("WEBP"))
                        this->Files["WEBP"].push_back(FileName);
                    else if (FileName.ends_with("webm") or FileName.ends_with("WEBM"))
                        this->Files["WEBM"].push_back(FileName);
                    else if (FileName.ends_with("heic") or FileName.ends_with("HEIC"))
                        this->Files["HEIC"].push_back(FileName);
                    else
                        this->Files["ETC"].push_back(FileName);
                }
            }
            return true;
        }

        catch (const std::exception &e) {
            ui->Logs->append("Error: " + QString(e.what()));
        }
    }
}

void FileOrganizer::Warning(QString Format, int Case) {
    switch (Case) {
    case 1:
        ui->Logs->append("Warning: There is no " + Format + " file or your request is invalid.");
        break;

    case 2:
        ui->Logs->append("Warning: This directory does not exists !");
        break;

    case 3:
        ui->Logs->append("Warning: This directory is empty !");
        break;

    default:
        ui->Logs->append("Warning: Incorrect input !");
        break;
    }
}

void FileOrganizer::on_Show_clicked() {
    ui->Logs->clear();
    ui->Logs->append("\t\t   *** Show Directory ***\n");

    if (this->SaveFilesNames()) {
        for (const auto& pair : this->Files) {
            if (!pair.second.empty()) {
                int index{ 0 };
                ui->Logs->append(QString::fromStdString(pair.first));
                for (const auto& value : pair.second) {
                    ui->Logs->append('[' + QString::number(++index) +
                                     '/' + QString::number(pair.second.size()) +
                                     "] -- " + QString::fromStdString(value));
                }
                ui->Logs->append("");
            }
        }
    }
}

void FileOrganizer::on_Categories_clicked() {
    ui->Logs->clear();
    ui->Logs->append("\t\t    *** Categories ***\n");

    if (this->SaveFilesNames()) {
        ui->Logs->append("New Directoires:");
        for (const auto& pair : this->Files) {
            if (!pair.second.empty()) {
                str NewPath{ this->Path + pair.first + '/' };
                if (!fs::exists(NewPath)) {
                    fs::create_directory(NewPath);

                    for (const auto& value : pair.second) {
                        fs::rename(this->Path + value, NewPath + value);
                    }

                    ui->Logs->append("+ " + QString::fromStdString(pair.first) +
                                     "[+" + QString::number(pair.second.size()) +
                                     " Files]");
                }
            }
        }
    }
}

void FileOrganizer::on_Rename_clicked() {
    ui->Logs->clear();
    ui->Logs->append("\t\t      *** Rename ***\n");

    if (this->SaveFilesNames()) {
        bool flag{ false };

        for (const auto& pair : this->Files) {
            if (QString::fromStdString(pair.first) == ui->RenameComboBox->currentText()) {
                int index{ 0 };
                for (const auto& value : pair.second) {
                    fs::rename(this->Path + value, this->Path + std::to_string(++index) + "." + pair.first);
                    ui->Logs->append("- " + QString::fromStdString(value));
                }

                ui->Logs->append("\nResult: "
                                 + QString::number(pair.second.size()) + ' '
                                 + QString::fromStdString(pair.first)
                                 + " files have been successfully renamed.");

                flag = true;
                break;
            }
        }

        if (!flag) {
            this->Warning(ui->RenameComboBox->currentText(), 1);
        }
    }
}

void FileOrganizer::on_Delete_clicked() {
    ui->Logs->clear();
    ui->Logs->append("\t\t      *** Delete ***\n");

    if (this->SaveFilesNames()) {
        bool flag{ false };

        if (ui->DeleteComboBox->currentText() == "ALL") {
            for (const auto& pair : this->Files) {
                for (const auto& value : pair.second) {
                    if (fs::remove(this->Path + value)) {
                        ui->Logs->append("- " + QString::fromStdString(value));
                    }
                }
            }

            ui->Logs->append("\nResult: All files have been successfully deleted.");
            flag = true;
        }

        for (const auto& pair : this->Files) {
            if (QString::fromStdString(pair.first) == ui->DeleteComboBox->currentText()) {
                for (const auto& value : pair.second) {
                    if (fs::remove(this->Path + value)) {
                        ui->Logs->append("- " + QString::fromStdString(value));
                    }
                }

                ui->Logs->append("\nResult: "
                                 + QString::number(pair.second.size()) + ' '
                                 + QString::fromStdString(pair.first)
                                 + " files have been successfully deleted.");

                flag = true;
                break;
            }
        }

        if (!flag) {
            this->Warning(ui->DeleteComboBox->currentText(), 1);
        }
    }
}
